File: Mis últimos pasos/BackEnd/SoftPetBusiness/src/main/java/pe/edu/pucp/softpet/bo/UsuarioBo.java
package pe.edu.pucp.softpet.bo;

import java.util.ArrayList;
import pe.edu.pucp.softpet.bo.utils.GmailService;
import pe.edu.pucp.softpet.dao.UsuarioDAO;
import pe.edu.pucp.softpet.daoImp.UsuarioDaoImpl;
import pe.edu.pucp.softpet.bo.utils.SecurityUtil;
import pe.edu.pucp.softpet.dto.usuarios.UsuarioDto;

public class UsuarioBo {

    private final UsuarioDAO usuarioDao;

    // Necesitamos GmailService aquí también si no lo tenías instanciado
    private final GmailService gmailService; 
    
    public UsuarioBo() {
        this.usuarioDao = new UsuarioDaoImpl();
        
        this.gmailService = new GmailService();
    }

    public Integer insertar(String username, String password,
            String correo, boolean activo) {

        UsuarioDto usuarioDto = new UsuarioDto();

        usuarioDto.setUsername(username);
        usuarioDto.setPassword(password);
        usuarioDto.setCorreo(correo);
        usuarioDto.setActivo(activo);

        return this.usuarioDao.insertar(usuarioDto);
    }

    public Integer modificar(int usuarioId, String username, String password,
            String correo, boolean activo) {

        UsuarioDto usuarioDto = new UsuarioDto();

        usuarioDto.setUsuarioId(usuarioId);
        usuarioDto.setUsername(username);
        usuarioDto.setPassword(password);
        usuarioDto.setCorreo(correo);
        usuarioDto.setActivo(activo);

        return this.usuarioDao.modificar(usuarioDto);
    }

    public Integer eliminar(Integer usuarioId) {
        UsuarioDto usuarioDto = new UsuarioDto();
        usuarioDto.setUsuarioId(usuarioId);
        return this.usuarioDao.eliminar(usuarioDto);
    }

    public UsuarioDto obtenerPorId(Integer usuarioId) {
        UsuarioDto usuarioDto = new UsuarioDto();
        usuarioDto.setUsuarioId(usuarioId);
        return this.usuarioDao.obtenerPorId(usuarioId);
    }

    public ArrayList<UsuarioDto> listarTodos() {
        return this.usuarioDao.listarTodos();
    }

//    public ArrayList<UsuarioDto> ObtenerPorCorreoYContra(String correo, String contra) {
//        return this.usuarioDao.ObtenerPorCorreoYContra(correo, contra);
//    }
    
    // MÉTODO MODIFICADO: Login seguro con Hashing
    public ArrayList<UsuarioDto> ObtenerPorCorreoYContra(String correo, String contraRaw) {
        // 1. El usuario ingresa "123456"
        // 2. Nosotros lo convertimos a Hash SHA-256
        String contraHash = SecurityUtil.sha256(contraRaw);
        
        // 3. Consultamos a la BD comparando HASH vs HASH
        return this.usuarioDao.ObtenerPorCorreoYContra(correo, contraHash);
    }

    // MÉTODO NUEVO: Cambiar Contraseña
    public Integer cambiarPassword(int idUsuario, String passwordActualRaw, String passwordNuevaRaw) {
        // 1. Obtener Hash almacenado en BD
        String currentHashDB = usuarioDao.obtenerPasswordActual(idUsuario);
        
        if (currentHashDB == null) return 0; // Usuario no existe

        // 2. Hashear la contraseña que ingresó el usuario para validar
        String inputHash = SecurityUtil.sha256(passwordActualRaw);
        
        if (currentHashDB.equals(inputHash)) {
            // 3. Si coincide, hashear la nueva y actualizar
            String newHash = SecurityUtil.sha256(passwordNuevaRaw);
            return usuarioDao.actualizarPassword(idUsuario, newHash);
        } else {
            return -1; // Contraseña actual incorrecta
        }
    }
    
    // NUEVO: Recuperar Contraseña
    public Integer recuperarPassword(String correo) {
        // 1. Verificar si el correo existe en BD
        UsuarioDto usuario = usuarioDao.obtenerPorCorreo(correo);
        
        if (usuario == null) {
            return -1; // Correo no registrado o inactivo
        }

        // 2. Generar nueva contraseña temporal
        String passTemporal = SecurityUtil.generarPasswordAleatoria(); // Ej: "Xy9Za2"
        
        // 3. Hashear para guardar en BD
        String passHash = SecurityUtil.sha256(passTemporal);
        
        // 4. Actualizar en BD
        int resultado = usuarioDao.actualizarPassword(usuario.getUsuarioId(), passHash);
        
        if (resultado > 0) {
            // 5. Enviar correo con la clave TEMPORAL (Raw)
            // Reutilizamos enviarCorreo_Credenciales que ya tienes configurado
            new Thread(() -> {
                gmailService.enviarCorreo_Credenciales(correo, passTemporal);
            }).start();
        }
        
        return resultado;
    }
}

File: Mis últimos pasos/SoftPetWA/SoftPetBussiness/SOAPBo/Seguridad/UsuarioBo.cs
using SoftPetBussiness.UsuarioClient;
using System.Collections.Generic;
using System.Linq;

namespace SoftPetBusiness
{
    public class UsuarioBO
    {
        private UsuariosClient clienteSOAP;

        public UsuarioBO()
        {
            this.clienteSOAP = new UsuariosClient();
        }

        // Insertar usuario
        public int Insertar(string username, string password, string correo, bool activo)
        {
            return this.clienteSOAP.insertar(username, password, correo, activo);
        }

        // Modificar usuario
        public int Modificar(int usuarioId, string username, string password, string correo, bool activo)
        {
            return this.clienteSOAP.modificar(usuarioId, username, password, correo, activo);
        }

        // Eliminar usuario
        public int Eliminar(int usuarioId)
        {
            return this.clienteSOAP.eliminar(usuarioId);
        }

        // Obtener usuario por ID
        public usuarioDto ObtenerPorId(int usuarioId)
        {
            return this.clienteSOAP.obtenerPorId(usuarioId);
        }

        // Listar todos los usuarios
        public List<usuarioDto> ListarTodos()
        {
            return this.clienteSOAP.listarTodos().ToList<usuarioDto>();
        }

        // Insertar usando DTO
        public int Insertar(usuarioDto usuario)
        {
            return this.clienteSOAP.insertar(
                usuario.username,
                usuario.password,
                usuario.correo,
                usuario.activo
            );
        }

        // Modificar usando DTO
        public int Modificar(usuarioDto usuario)
        {
            return this.clienteSOAP.modificar(
                usuario.usuarioId,
                usuario.username,
                usuario.password,
                usuario.correo,
                usuario.activo
            );
        }

        // Login Seguro (Hash)
        public List<usuarioDto> ObtenerPorCorreoYContra(string correo, string contra)
        {
            // 1. Llama al servicio SOAP (que ahora espera la contraseña en texto plano para hashearla)
            // O si tu lógica cambió para hashear aquí, asegúrate de coordinarlo. 
            // Según lo que hicimos en Java, el BO de Java hashea. Así que aquí enviamos plano.
            var respuestaSOAP = this.clienteSOAP.ObtenerPorCorreoYContra(correo, contra);

            // 2. Comprueba si la respuesta es nula
            if (respuestaSOAP == null)
            {
                return new List<usuarioDto>();
            }
            else
            {
                return respuestaSOAP.ToList<usuarioDto>();
            }
        }

        // NUEVO MÉTODO: Cambiar Contraseña
        public int cambiarPassword(int idUsuario, string passwordActual, string passwordNueva)
        {
            // Llama al nuevo método expuesto en el WS Java
            // Si te sale error aquí, haz clic derecho en 'Connected Services' -> 'Update Service Reference'
            return this.clienteSOAP.cambiarPassword(idUsuario, passwordActual, passwordNueva);
        }

        // ========================================================
        // NUEVO MÉTODO: RECUPERAR CONTRASEÑA
        // ========================================================
        public int recuperarPassword(string correo)
        {
            // Llama al servicio Java que se encarga de todo el proceso seguro
            return this.clienteSOAP.recuperarPassword(correo);
        }
    }
}